src/twisted/internet/cfreactor.py

```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
         self._scheduleSimulate()
         return result
 
+    def startRunning(self, installSignalHandlers: bool = True) -> None:
+        """
+        Start running the reactor, then kick off the timer that advances
+        Twisted's clock to keep pace with CFRunLoop's.
+        """
+        super().startRunning(installSignalHandlers)
+        # We must force the simulator to run because in addition to scheduling
+        # timed calls, super().startRunning also fires startup events which may
+        # crash the reactor.
+        self._scheduleSimulate(force=True)
+
     _inCFLoop = False
 
-    def mainLoop(self):
+    def mainLoop(self) -> None:
         """
         Run the runner (C{CFRunLoopRun} or something that calls it), which runs
         the run loop until C{crash()} is called.
         """
-        self._inCFLoop = True
+        # we were crashed during startup.
+        if not self._started:
+
+            def docrash() -> None:
+                self.crash()
+
+            # we're running again, just for a moment, while we wait for some
+            # I/O to take place.  We need this to be True for callLater to not
+            # be a functional no-op.  the crash timer will set us back to
+            # _started = False again momentarily.
+            self._started = True
+            self.callLater(0, docrash)
+        already = False
         try:
-            self._runner()
+            while self._started:
+                if already:
+                    # Sometimes CFRunLoopRun (or its equivalents) may exit
+                    # early. This is really only *supposed* to happen when it
+                    # runs out of sources & timers to process, but it appears
+                    # that in cases where the reactor is being rapidly started
+                    # and crashed repeatedly, and sources are being rapidly
+                    # invalidated both inside and outside the loop (e.g. in the
+                    # Twisted test suite), CoreFoundation (possibly
+                    # incorrectly?) exits the runloop early.  To accommodate
+                    # this stress case, we make extra sure that our scheduled
+                    # timer is re-created on the loop as a CFRunLoopTimer.
```

Comment:
```suggestion
                    # Twisted test suite), CoreFoundation exits the runloop even
                    # when there *is* some work to do. To accommodate this
                    # stress case, we make extra sure that our scheduled
                    # timer is re-created on the loop as a CFRunLoopTimer.
```

Reply:
I cannot prove that there is in fact "work to do", at least according to the semantics established by CFRunLoop: there are various points where sources are automatically invalidated (when timers expire, I think), and some of our uses of the API (for example, `CFSocketCreateWithNative` with a pipe file descriptor rather than a socket) are taking some liberties. So it's possible that we just haven't correctly appreciated some nuance of automatic invalidation

Reply:
Much lengthier comment hopefully explains the situation in detail.